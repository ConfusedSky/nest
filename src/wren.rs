#![allow(unsafe_code)]

use std::{
    borrow::Cow,
    cell::RefCell,
    ffi::{c_void, CStr, CString},
    mem::MaybeUninit,
    os::raw,
    pin::Pin,
    ptr::NonNull,
};

use crate::wren_sys::{
    self, wrenCall, wrenFreeVM, wrenGetUserData, wrenGetVariable, wrenInitConfiguration,
    wrenInterpret, wrenMakeCallHandle, wrenNewVM, WrenConfiguration, WrenErrorType, WrenHandle,
    WrenInterpretResult, WrenVM,
};

unsafe fn get_user_data<'s, V>(vm: *mut WrenVM) -> Option<&'s mut V> {
    let user_data = wrenGetUserData(vm);
    if user_data.is_null() {
        None
    } else {
        Some(user_data.cast::<V>().as_mut().unwrap())
    }
}

unsafe extern "C" fn write_fn<V: VmUserData>(vm: *mut WrenVM, text: *const i8) {
    let user_data = get_user_data::<V>(vm);

    if let Some(user_data) = user_data {
        let text = CStr::from_ptr(text).to_string_lossy();
        user_data.on_write(text.as_ref());
    }
}

unsafe extern "C" fn error_fn<V: VmUserData>(
    vm: *mut WrenVM,
    error_type: WrenErrorType,
    module: *const i8,
    line: i32,
    msg: *const i8,
) {
    let user_data = get_user_data::<V>(vm);
    if let Some(user_data) = user_data {
        let msg = CStr::from_ptr(msg).to_string_lossy();
        // This lives outside of the if statement so that it can live long enough
        // to be passed to user_data on error
        let c_module: Cow<str>;
        // Runtime doesn't have a valid module so it will crash if it goes any further
        let kind = if error_type == wren_sys::WrenErrorType_WREN_ERROR_RUNTIME {
            ErrorKind::Runtime(msg.as_ref())
        } else {
            c_module = CStr::from_ptr(module).to_string_lossy();
            let context = ErrorContext {
                module: c_module.as_ref(),
                line,
                msg: msg.as_ref(),
            };
            match error_type {
                wren_sys::WrenErrorType_WREN_ERROR_COMPILE => ErrorKind::Compile(context),
                wren_sys::WrenErrorType_WREN_ERROR_RUNTIME => ErrorKind::Runtime(msg.as_ref()),
                wren_sys::WrenErrorType_WREN_ERROR_STACK_TRACE => ErrorKind::Stacktrace(context),
                kind => ErrorKind::Unknown(kind, context),
            }
        };

        user_data.on_error(kind);
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy)]
pub struct VMPtr(NonNull<WrenVM>);

type Slot = std::os::raw::c_int;

impl VMPtr {
    /// SAFETY: Will segfault if an invalid slot
    /// is asked for
    pub unsafe fn set_slot_bool_unchecked(self, slot: Slot, value: bool) {
        wren_sys::wrenSetSlotBool(self.0.as_ptr(), slot, value);
    }

    /// SAFETY: Calling this on a slot that isn't a bool or a valid slot is undefined behavior
    pub unsafe fn get_slot_bool_unchecked(self, slot: Slot) -> bool {
        wren_sys::wrenGetSlotBool(self.0.as_ptr(), slot)
    }

    /// SAFETY: this is always non null but will segfault if an invalid slot
    /// is asked for
    /// And is not guarenteed to be a valid object
    pub unsafe fn get_slot_handle_unchecked(self, slot: Slot) -> NonNull<WrenHandle> {
        NonNull::new_unchecked(wren_sys::wrenGetSlotHandle(self.0.as_ptr(), slot))
    }

    /// SAFETY: this is always non null but will segfault if an invalid slot
    /// is asked for
    /// MAYBE: Will seg fault if the variable does not exist?
    /// Still need to set up module resolution
    pub unsafe fn get_variable_unchecked<Module, Name>(self, module: Module, name: Name, slot: Slot)
    where
        Module: AsRef<str>,
        Name: AsRef<str>,
    {
        let vm = self.0;
        let module = CString::new(module.as_ref()).unwrap();
        let name = CString::new(name.as_ref()).unwrap();

        wrenGetVariable(vm.as_ptr(), module.as_ptr(), name.as_ptr(), slot);
    }

    pub fn make_call_handle<Signature>(self, signature: Signature) -> NonNull<WrenHandle>
    where
        Signature: AsRef<str>,
    {
        let vm = self.0;
        let signature = CString::new(signature.as_ref()).unwrap();
        // SAFETY: this function is always safe to call but may be unsafe to use the handle it returns
        // as that handle might not be valid
        unsafe { NonNull::new_unchecked(wrenMakeCallHandle(vm.as_ptr(), signature.as_ptr())) }
    }

    /// Safety: Will segfault if used with an invalid method
    pub unsafe fn call(self, method: NonNull<WrenHandle>) -> Result<(), InterpretResultErrorKind> {
        let vm = self.0;
        let result = wrenCall(vm.as_ptr(), method.as_ptr());

        InterpretResultErrorKind::new_from_result(result)
    }

    pub fn ensure_slots(self, num_slots: Slot) {
        // SAFETY: this one is always safe to call even if the value is negative
        unsafe {
            wren_sys::wrenEnsureSlots(self.0.as_ptr(), num_slots);
        }
    }
}

pub struct ErrorContext<'s> {
    pub module: &'s str,
    pub line: i32,
    pub msg: &'s str,
}

pub enum ErrorKind<'s> {
    Compile(ErrorContext<'s>),
    Runtime(&'s str),
    Stacktrace(ErrorContext<'s>),
    Unknown(WrenErrorType, ErrorContext<'s>),
}

pub enum InterpretResultErrorKind {
    Compile,
    Runtime,
    Unknown(WrenInterpretResult),
}

impl InterpretResultErrorKind {
    const fn new_from_result(result: u32) -> Result<(), Self> {
        match result {
            wren_sys::WrenInterpretResult_WREN_RESULT_COMPILE_ERROR => Err(Self::Compile),
            wren_sys::WrenInterpretResult_WREN_RESULT_RUNTIME_ERROR => Err(Self::Runtime),
            wren_sys::WrenInterpretResult_WREN_RESULT_SUCCESS => Ok(()),
            kind => Err(Self::Unknown(kind)),
        }
    }
}

#[allow(unused_variables)]
// We define empty defaults here so that the user can define what they want
pub trait VmUserData {
    fn on_write(&mut self, text: &str) {}
    fn on_error(&mut self, kind: ErrorKind) {}
}

pub struct Vm<V> {
    vm: VMPtr,
    // This value is held here so that it is
    // disposed of properly when execution is finished
    // but it isn't actually used in the struct
    _user_data: Pin<Box<RefCell<V>>>,
}

impl<V> Drop for Vm<V> {
    fn drop(&mut self) {
        unsafe { wrenFreeVM(self.vm.0.as_ptr()) }
    }
}

impl<V> Vm<V>
where
    V: VmUserData,
{
    pub fn new(user_data: V) -> Option<Self> {
        unsafe {
            let mut config: WrenConfiguration = MaybeUninit::zeroed().assume_init();
            wrenInitConfiguration(&mut config);

            let user_data = Box::pin(RefCell::new(user_data));

            config.writeFn = Some(write_fn::<V>);
            config.errorFn = Some(error_fn::<V>);
            config.userData = user_data.as_ptr().cast::<c_void>();

            let vm = VMPtr(NonNull::new(wrenNewVM(&mut config))?);

            Some(Self {
                vm,
                _user_data: user_data,
            })
        }
    }

    pub fn interpret<M, S>(&self, module: M, source: S) -> Result<(), InterpretResultErrorKind>
    where
        M: AsRef<str>,
        S: AsRef<str>,
    {
        unsafe {
            let module = CString::new(module.as_ref()).unwrap();
            let source = CString::new(source.as_ref()).unwrap();
            let result = wrenInterpret(self.vm.0.as_ptr(), module.as_ptr(), source.as_ptr());

            InterpretResultErrorKind::new_from_result(result)
        }
    }
}
