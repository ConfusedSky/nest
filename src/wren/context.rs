use std::{
    ffi::{CStr, CString, FromBytesWithNulError},
    marker::PhantomData,
    ops::{Deref, DerefMut},
    ptr::NonNull,
};

use wren_sys::{self as ffi, WrenVM};

use super::{
    foreign, handle::CallHandle, system_methods::SystemMethods, value::TryGetResult, Fiber, Get,
    GetArgs, Handle, InterpretResultErrorKind, Result, Set, SetArgs, Slot, SystemUserData, Value,
    VmUserData,
};

pub type Raw<'wren, L> = Context<'wren, NoTypeInfo, L>;
pub type RawForeign<'wren> = Raw<'wren, Foreign>;
pub type RawNative<'wren> = Raw<'wren, Native>;

#[repr(transparent)]
#[derive(Debug, Clone)]
pub struct Context<'wren, T, L: Location>(
    NonNull<WrenVM>,
    PhantomData<&'wren mut SystemUserData<'wren, T>>,
    PhantomData<L>,
);

impl<'wren, V, L: Location> Context<'wren, V, L> {
    const unsafe fn transmute<V2, L2: Location>(&self) -> &Context<'wren, V2, L2> {
        &*((self as *const Context<V, L>).cast::<Context<V2, L2>>())
    }
    unsafe fn transmute_mut<V2, L2: Location>(&mut self) -> &mut Context<'wren, V2, L2> {
        &mut *((self as *mut Context<V, L>).cast::<Context<V2, L2>>())
    }

    // NOTE THESE ARE ALL DOWNCASTS SO THIS IS SAFE

    pub const fn as_foreign(&self) -> &Context<'wren, V, Foreign> {
        unsafe { self.transmute::<V, Foreign>() }
    }

    pub fn as_foreign_mut(&mut self) -> &mut Context<'wren, V, Foreign> {
        unsafe { self.transmute_mut::<V, Foreign>() }
    }

    pub const fn as_raw(&self) -> &Context<'wren, NoTypeInfo, L> {
        unsafe { self.transmute::<NoTypeInfo, L>() }
    }
    pub fn as_raw_mut(&mut self) -> &mut Context<'wren, NoTypeInfo, L> {
        unsafe { self.transmute_mut::<NoTypeInfo, L>() }
    }

    pub(super) const fn as_ptr(&self) -> *mut WrenVM {
        self.0.as_ptr()
    }

    #[allow(dead_code)]
    pub(super) unsafe fn new(vm: *mut WrenVM) -> Option<Self> {
        Some(Self(NonNull::new(vm)?, PhantomData, PhantomData))
    }

    pub(super) unsafe fn new_unchecked(vm: *mut WrenVM) -> Self {
        Self(NonNull::new_unchecked(vm), PhantomData, PhantomData)
    }
}

// Type information is needed to get user data
impl<'wren, V: VmUserData<'wren, V>, L: Location> Context<'wren, V, L> {
    pub fn get_user_data(&self) -> &V {
        // SAFETY this is called from a typed context
        unsafe { &foreign::get_system_user_data(self.as_ptr()).user_data }
    }
    pub fn get_user_data_mut(&mut self) -> &mut V {
        // SAFETY this is called from a typed context
        unsafe { &mut foreign::get_system_user_data(self.as_ptr()).user_data }
    }
    pub fn get_user_data_mut_with_context(
        &mut self,
    ) -> (&mut V, &mut Context<'wren, NoTypeInfo, L>) {
        unsafe {
            (
                &mut foreign::get_system_user_data(self.0.as_ptr()).user_data,
                self.as_raw_mut(),
            )
        }
    }
}

impl<'wren, V: VmUserData<'wren, V>> Deref for Context<'wren, V, Native> {
    type Target = RawNative<'wren>;
    fn deref(&self) -> &Self::Target {
        self.as_raw()
    }
}
impl<'wren, V: VmUserData<'wren, V>> DerefMut for Context<'wren, V, Native> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.as_raw_mut()
    }
}

impl<'wren, V: VmUserData<'wren, V>> Deref for Context<'wren, V, Foreign> {
    type Target = RawForeign<'wren>;
    fn deref(&self) -> &Self::Target {
        self.as_raw()
    }
}
impl<'wren, V: VmUserData<'wren, V>> DerefMut for Context<'wren, V, Foreign> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.as_raw_mut()
    }
}

// Calling can only happen from a native context
impl<'wren, T> Context<'wren, T, Native> {
    pub fn interpret<M, S>(&self, module: M, source: S) -> Result<()>
    where
        M: AsRef<str>,
        S: AsRef<str>,
    {
        // SAFETY this will segfault if it is called within a foreign function
        // so it's safe because it must be used in a native context
        unsafe {
            let module = CString::new(module.as_ref()).unwrap();
            let source = CString::new(source.as_ref()).unwrap();
            let result = ffi::wrenInterpret(self.as_ptr(), module.as_ptr(), source.as_ptr());

            InterpretResultErrorKind::new_from_result(result)
        }
    }

    /// Call [method] on a [subject] with [args] on the vm
    /// subject is usually a class or an object, but all calls require a subject
    /// this is safe as long as the return type specified is correct
    /// otherwise it's UB
    /// Arguments must be set up correctly as well
    pub unsafe fn call_unchecked<G: Get<'wren, Native>, Args: SetArgs<'wren, Native>>(
        &mut self,
        subject: &Handle<'wren>,
        method: &CallHandle<'wren>,
        args: &Args,
    ) -> Result<G> {
        // make sure there enough slots to send over the subject
        // and all it's args
        let vm = self.as_raw_mut();
        vm.ensure_slots(1 + Handle::ADDITIONAL_SLOTS_NEEDED + Args::REQUIRED_SLOTS);

        // Send over the subject and all of it's args
        subject.send_to_vm(vm, 0);
        args.set_slots(vm, 1);

        let result = ffi::wrenCall(vm.as_ptr(), method.as_ptr());
        InterpretResultErrorKind::new_from_result(result)?;

        // This should be safe as long as the type is set correctly
        // TODO: Make this safer probably use a try_get
        Ok(vm.get_return_value_unchecked::<G>())
    }

    /// Checks a handle to see if it is a valid fiber, if it is return the handle as a fiber
    /// in the ok varient, otherwise returns the original handle
    pub fn check_fiber(&mut self, handle: Handle<'wren>) -> TryGetResult<'wren, Fiber<'wren>> {
        let vm = self.as_raw_mut();
        vm.get_system_methods().fiber_methods.construct(vm, handle)
    }
}

impl<'wren, L: Location> Context<'wren, NoTypeInfo, L> {
    pub(super) fn get_system_methods<'s>(&self) -> &'s SystemMethods<'wren> {
        unsafe {
            foreign::get_system_user_data::<()>(self.as_ptr())
                .system_methods
                .as_ref()
                .expect("SystemMethods should be initialized at this point")
        }
    }

    pub fn get_variable<Module, Name>(
        &mut self,
        module: Module,
        name: Name,
        slot: Slot,
    ) -> Option<Handle<'wren>>
    where
        Module: AsRef<str>,
        Name: AsRef<str>,
    {
        let module = CString::new(module.as_ref()).unwrap();
        let name = CString::new(name.as_ref()).unwrap();
        // SAFETY wrenGetVariable is definitely safe if wrenHasModule and wrenHasVariable
        // are called beforehand
        // wrenHasVariable is safe if wrenHasModule has been called
        // and wrenHasModule is always safe to call
        unsafe {
            if !ffi::wrenHasModule(self.as_ptr(), module.as_ptr())
                || !ffi::wrenHasVariable(self.as_ptr(), module.as_ptr(), name.as_ptr())
            {
                None
            } else {
                Some(self.get_variable_unchecked(module.as_c_str(), name.as_c_str(), slot))
            }
        }
    }

    /// SAFETY: this is always non null but will segfault if an invalid slot
    /// is asked for
    /// MAYBE: Will seg fault if the variable does not exist?
    /// Still need to set up module resolution
    pub unsafe fn get_variable_unchecked(
        &mut self,
        module: &CStr,
        name: &CStr,
        slot: Slot,
    ) -> Handle<'wren> {
        ffi::wrenGetVariable(self.as_ptr(), module.as_ptr(), name.as_ptr(), slot);
        Handle::get_from_vm(self.as_foreign_mut(), slot)
    }

    pub fn make_call_handle_slice(
        &mut self,
        signature: &[u8],
    ) -> std::result::Result<CallHandle<'wren>, FromBytesWithNulError> {
        CallHandle::new_from_slice(self.as_foreign_mut(), signature)
    }

    pub fn make_call_handle(&mut self, signature: &CStr) -> CallHandle<'wren> {
        CallHandle::new_from_signature(self.as_foreign_mut(), signature)
    }

    pub fn ensure_slots(&mut self, num_slots: Slot) {
        // SAFETY: this one is always safe to call even if the value is negative
        unsafe {
            wren_sys::wrenEnsureSlots(self.as_ptr(), num_slots);
        }
    }

    pub fn set_stack<Args: SetArgs<'wren, L>>(&mut self, args: &Args) {
        args.set_wren_stack(self, 0);
    }

    pub fn set_return_value<Args: Set<'wren, L> + ?Sized>(&mut self, arg: &Args) {
        arg.set_wren_stack(self, 0);
    }

    // TODO: Create safe version that returns Options depending on how many slots
    // there are
    pub unsafe fn get_stack_unchecked<Args: GetArgs<'wren, L>>(&mut self) -> Args {
        Args::get_slots(self)
    }

    pub unsafe fn get_return_value_unchecked<Args: Get<'wren, L>>(&mut self) -> Args {
        Args::get_slots(self)
    }

    // TODO: Make sure this can only be run in a foreign context
    // since it requires a fiber to be active in the vm
    pub fn abort_fiber<S>(&mut self, value: S)
    where
        S: AsRef<str>,
    {
        self.set_return_value(&value.as_ref());
        unsafe {
            ffi::wrenAbortFiber(self.as_ptr(), 0);
        }
    }
}

#[derive(Clone)]
pub struct NoTypeInfo;

mod sealed {
    use super::{Foreign, Native};

    pub trait Location {}
    impl Location for Foreign {}
    impl Location for Native {}
}

pub trait Location: sealed::Location {}
#[derive(Clone)]
pub struct Foreign;
impl Location for Foreign {}
#[derive(Clone)]
pub struct Native;
impl Location for Native {}

mod assert {
    use super::{Context, Native, VmUserData, WrenVM};

    struct T;
    impl<'wren> VmUserData<'wren, Self> for T {}
    // Ensure that VMPtr is the same Size as `*mut WrenVM`
    // the whole purpose of it is to make it easier to access
    // the wren api, without having to sacrifice size, performance or ergonomics
    // So they should be directly castable
    static_assertions::assert_eq_align!(Context<T, Native>, *mut WrenVM);
    static_assertions::assert_eq_size!(Context<T, Native>, *mut WrenVM);
}
