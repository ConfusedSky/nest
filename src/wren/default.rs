pub fn on_error(kind: super::ErrorKind) {
    match kind {
        super::ErrorKind::Compile(ctx) => {
            println!("[{} line {}] [Error] {}", ctx.module, ctx.line, ctx.msg);
        }
        super::ErrorKind::Runtime(msg) => println!("[Runtime Error] {}", msg),
        super::ErrorKind::Stacktrace(ctx) => {
            println!("[{} line {}] in {}", ctx.module, ctx.line, ctx.msg);
        }
        super::ErrorKind::Unknown(kind, ctx) => {
            println!(
                "[{} line {}] [Unkown Error {}] {}",
                ctx.module, ctx.line, kind, ctx.msg
            );
        }
    }
}
