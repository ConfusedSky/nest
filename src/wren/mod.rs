#![allow(unsafe_code)]

pub mod context;
mod fiber;
mod foreign;
mod handle;
mod system_methods;
#[cfg(test)]
mod test;
pub mod user_data;
mod value;

pub use fiber::Fiber;
pub use handle::{CallHandle, Handle};
pub use user_data::UserData as VmUserData;
pub use value::{GetArgs, GetValue, SetArgs, SetValue};

pub use wren_sys::WREN_VERSION_STRING as VERSION;

use std::{ffi::c_void, marker::PhantomData, mem::ManuallyDrop};

pub use self::{
    context::{Context, RawForeign as RawForeignContext, RawNative as RawNativeContext},
    system_methods::SystemMethods,
};
use wren_sys::{self as ffi, WrenErrorType, WrenInterpretResult, WrenVM};

pub type ForeignMethod<'wren, T> = unsafe fn(vm: Context<'wren, T, context::Foreign>);
pub type Result<T> = std::result::Result<T, InterpretResultErrorKind>;

#[macro_export]
macro_rules! cstr {
    ($s:expr) => {{
        use std::ffi::CStr;
        const CSTR: *const i8 =
            (concat!($s, "\0") as *const str as *const [::std::os::raw::c_char]).cast::<i8>();
        #[allow(unused_unsafe)]
        unsafe {
            CStr::from_ptr(CSTR)
        }
    }};
}
pub use cstr;

pub type Slot = std::os::raw::c_int;

#[derive(Debug)]
pub struct ErrorContext<'s> {
    pub module: &'s str,
    pub line: i32,
    pub msg: &'s str,
}

#[derive(Debug)]
pub enum ErrorKind<'s> {
    Compile(ErrorContext<'s>),
    Runtime(&'s str),
    Stacktrace(ErrorContext<'s>),
    Unknown(WrenErrorType, ErrorContext<'s>),
}

#[derive(Debug)]
pub enum InterpretResultErrorKind {
    Compile,
    Runtime,
    IncorrectNumberOfArgsPassed,
    Unknown(WrenInterpretResult),
}

impl InterpretResultErrorKind {
    const fn new_from_result(result: u32) -> Result<()> {
        match result {
            wren_sys::WrenInterpretResult_WREN_RESULT_COMPILE_ERROR => Err(Self::Compile),
            wren_sys::WrenInterpretResult_WREN_RESULT_RUNTIME_ERROR => Err(Self::Runtime),
            wren_sys::WrenInterpretResult_WREN_RESULT_SUCCESS => Ok(()),
            kind => Err(Self::Unknown(kind)),
        }
    }
}

// The values contained in here are boxed because for some reason
// we see failures otherwise
// There is probably a more efficient way to implement this
#[repr(C)]
struct SystemUserData<'wren, V: 'wren> {
    system_methods: Option<SystemMethods<'wren>>,
    // User data must always be the last item in the struct because it is variable
    // size and sometimes we need to access other system user data items from an untyped
    // context so we want to make sure that while this can grow and shrink that it doesn't
    // affect the offsets to other items in the system data
    user_data: ManuallyDrop<V>,
}

impl<'wren, V> SystemUserData<'wren, V> {
    const fn new(user_data: V) -> Self {
        Self {
            user_data: ManuallyDrop::new(user_data),
            system_methods: None,
        }
    }
}

impl<'wren, V> Drop for SystemUserData<'wren, V> {
    fn drop(&mut self) {
        // Make sure user data is dropped first because there might be things that
        // the user_data depends on that are used in system methods
        unsafe {
            ManuallyDrop::drop(&mut self.user_data);
        }
        drop(self.system_methods.take());
    }
}

pub struct Vm<'wren, V: VmUserData<'wren, V>> {
    vm: Context<'wren, V, context::Native>,
    // The user data object is actually held and owned by the vm
    // We will handle dropping this data ourselves
    _phantom: PhantomData<SystemUserData<'wren, V>>,
}

impl<'wren, V> Drop for Vm<'wren, V>
where
    V: VmUserData<'wren, V>,
{
    fn drop(&mut self) {
        // Take and drop the user data here so that all handles are
        // freed before the vm is freed
        unsafe {
            let user_data = ffi::wrenGetUserData(self.vm.as_ptr());
            // Create a new box object and let it free itself before the
            // vm is freed
            drop(Box::<SystemUserData<V>>::from_raw(user_data.cast()));
            ffi::wrenFreeVM(self.as_ptr());
        }
    }
}

impl<'wren, V> Vm<'wren, V>
where
    V: VmUserData<'wren, V>,
{
    pub fn get_context(&mut self) -> &mut Context<'wren, V, context::Native> {
        &mut self.vm
    }

    const fn as_ptr(&self) -> *mut WrenVM {
        self.vm.as_ptr()
    }

    pub fn new(user_data: V) -> Self {
        unsafe {
            let mut config = foreign::init_config::<V>();

            let user_data = SystemUserData::new(user_data);
            let user_data = Box::new(user_data);
            let user_data_ptr = Box::into_raw(user_data);

            config.userData = user_data_ptr.cast::<c_void>();

            let mut vm: Context<V, context::Native> =
                Context::new_unchecked(ffi::wrenNewVM(&mut config));
            (*user_data_ptr).system_methods = Some(SystemMethods::new(&mut vm));

            Self {
                vm,
                _phantom: PhantomData,
            }
        }
    }
}
