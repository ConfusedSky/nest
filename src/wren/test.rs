use std::collections::HashMap;

use super::{context, Fiber, ForeignMethod, Handle, Vm, VmUserData};

pub type Context<'wren, L> = super::Context<'wren, UserData<'wren>, L>;

#[derive(Default)]
pub struct UserData<'wren> {
    output: String,
    static_foreign: HashMap<&'wren str, ForeignMethod<'wren, UserData<'wren>>>,
    pub fiber: Option<Fiber<'wren>>,
}

impl<'wren> VmUserData<'wren, Self> for UserData<'wren> {
    fn on_error(
        &mut self,
        _: super::Context<'wren, Self, context::Foreign>,
        kind: super::ErrorKind,
    ) {
        super::user_data::on_error(kind);
    }
    fn on_write(&mut self, _: super::Context<'wren, Self, context::Foreign>, text: &str) {
        print!("{}", text);
        self.output += text;
    }
    fn bind_foreign_method(
        &mut self,
        module: &str,
        classname: &str,
        is_static: bool,
        signature: &str,
    ) -> Option<ForeignMethod<'wren, Self>> {
        if module != "<test>" || !is_static || classname != "Test" {
            return None;
        }
        return self.static_foreign.get(signature).copied();
    }
}

impl<'wren> UserData<'wren> {
    pub const fn get_output(&self) -> &String {
        &self.output
    }

    pub fn set_static_foreign_method(
        &mut self,
        signature: &'wren str,
        foreign: ForeignMethod<'wren, UserData<'wren>>,
    ) {
        self.static_foreign.insert(signature, foreign);
    }
}

#[macro_export]
macro_rules! call_test_case {
        ($type:ty, $vm:ident { $class:ident.$handle:ident } == $res:expr) => {
            let slice = wren_macros::to_signature!($handle);
            let handle = $vm.make_call_handle_slice(slice).unwrap();
            let res: $type = crate::wren::util::make_call!($vm {$class.handle()}).expect(
                &format!(
                    "{}.{} is not a valid invocation",
                    stringify!($class),
                    stringify!($handle),
                )
            );
            assert_eq!( res, $res );
        };
        ($type:ty, $vm:ident { $class:ident.$handle:ident() } == $res:expr) => {
            let slice = wren_macros::to_signature!($handle());
            let handle = $vm.make_call_handle_slice(slice).unwrap();
            let res: $type = crate::wren::util::make_call!($vm {$class.handle()}).expect(
                &format!(
                    "{}.{} is not a valid invocation",
                    stringify!($class),
                    stringify!($handle),
                )
            );
            assert_eq!( res, $res );
        };
        ($type:ty, $vm:ident { $class:ident.$handle:ident($($args:expr),+ ) } == $res:expr) => {
            let slice = wren_macros::to_signature!($handle($($args),+ ));
            let handle = $vm.make_call_handle_slice(slice).unwrap();
            let res: $type = crate::wren::util::make_call!($vm { $class.handle($($args),+ ) }).expect(
                &format!(
                    "{}.{} is not a valid invocation",
                    stringify!($class),
                    stringify!($handle),
                )
            );
            assert_eq!( res, $res );
        };
    }

pub use call_test_case;

pub fn create_test_vm<'wren>(
    source: &str,
    fn_binding: impl FnOnce(&mut UserData<'wren>),
) -> (Vm<'wren, UserData<'wren>>, Handle<'wren>) {
    let mut vm = Vm::new(UserData::default());

    let vmptr = vm.get_context();
    fn_binding(vmptr.get_user_data_mut());
    vmptr
        .interpret("<test>", source)
        .expect("Code should run successfully");

    vmptr.ensure_slots(1);
    let class = vmptr
        .get_variable("<test>", "Test", 0)
        .expect("Test class should be defined in source");

    (vm, class)
}
