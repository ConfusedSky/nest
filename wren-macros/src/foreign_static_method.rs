//! # Goals
//! Note here the first checkbox here is implementation and the second one is tested
//!
//! - Create a wrapper around a function that makes get_slot calls to get
//!     the returned values [x] [x]
//!
//! - Make sure all foreign methods for a module are implemented at compile time
//!     IE: Have it be a compiler error if there are any foreign methods that haven't
//!     been implemented [ ] [ ]
//!
//! - Have the ability to optionally generate stub implementations that do some
//!     typechecking on the wren side for the public api of a class.
//!     Since we can't really do that on the rust side. [ ] [ ]
//!
//! - Have two modes strict and dynamic where depending on the type the function
//!     calls get_slot or get_slot_unchecked and the wren end user has to be more
//!     careful respectively [ ] [ ]
//!
//! - Optionally support results for the try_get methods [ ] [ ]
//!
//! - Allow the user to leave off the context in their arguments [ ] [ ]
//!   If they leave off the context from their arguments then the
//!     context user data should be a generic [ ] [ ]
//!   Otherwise it should be have the same type as the context passed in [ ] [ ]
//!   Make sure to check that it is a foreign context and error if it isn't [ ] [ ]
//!
//! - Generate better errors for bad return values
//!   Make sure the error appears at the return value [ ] [ ]
//!   Create a custom error message for a bad return value type [ ] [ ]
//!
//! - Make sure to respect visibility [ ] [ ]
//!
//! - Have good error messages
//!   Make sure the context is always the first item in the argument list [ ] [ ]
//!   Have errors saying which argument has an invalid type [ ] [ ]

use proc_macro2::{Ident, Span, TokenStream};
use proc_macro_crate::{crate_name, FoundCrate};
use quote::{quote, quote_spanned};
use syn::{spanned::Spanned, ItemFn, LitInt};

pub fn foreign_static_method(mut input: ItemFn) -> syn::Result<TokenStream> {
    let name = input.sig.ident;
    let original_function_name = quote::format_ident!("__wren_internal_{}", name);
    input.sig.ident = original_function_name.clone();

    let wren_crate = {
        let crate_name = crate_name("wren").expect("wren must be present for this macro");
        match crate_name {
            FoundCrate::Itself => Ident::new("crate", Span::call_site()),
            FoundCrate::Name(name) => Ident::new(&name, Span::call_site()),
        }
    };

    let args = input.sig.inputs.clone();
    let args = args
        .into_iter()
        .enumerate()
        .map(|(i, argument)| {
            if let syn::FnArg::Typed(pattern) = argument {
                let arg_name = pattern.pat.clone();
                let arg_type = &pattern.ty;
                // Start at 1 instead of 0 to make sure that we read the arguments
                // rather than the Class
                let i = LitInt::new(&(i + 1).to_string(), Span::call_site());

                let get_slot = quote_spanned!(
                    pattern.span() =>
                        let #arg_name =
                            #arg_type::get_slot(
                                &mut context,
                                #i
                            );
                );

                Ok((arg_name, get_slot))
            } else {
                Err(syn::Error::new(
                    argument.span(),
                    "This macro doesn't support instance methods",
                ))
            }
        })
        .collect::<syn::Result<Vec<_>>>()?;
    let arg_names = args.iter().map(|x| &x.0);
    let arg_get_slot = args.iter().map(|x| &x.1);

    Ok(quote!(
        #input

        unsafe fn #name<'wren, V: #wren_crate::VmUserData<'wren, V>>(
            mut context: #wren_crate::Context<
                'wren,
                V,
                #wren_crate::context::Foreign
            >
        ) {
            use #wren_crate::{GetValue, SetValue};
            #(#arg_get_slot)*
            #original_function_name(#(#arg_names),*)
                .set_slot(&mut context, 0);
        }
    ))
}
