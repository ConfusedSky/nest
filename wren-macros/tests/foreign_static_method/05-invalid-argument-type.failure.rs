//! This is the simplest happy path

use wren::test::{call_test_case, create_test_vm};
use wren_macros::foreign_static_method;

#[foreign_static_method]
fn test(a: f32, b: f32, c: f32) -> f32 {
    a + b + c
}

fn main() {
    let (mut vm, test) = create_test_vm(
        "class Test {
            foreign static foreignTest(a, b, c)
            static useForeignTest() { foreignTest(1, 2, 3) }
        }",
        |f| {
            f.set_static_foreign_method("foreignTest(_,_,_)", foreign_test);
        },
    );

    let context = vm.get_context();

    call_test_case!(context {
        test.useForeignTest() == Ok(6.0)
    });
}
