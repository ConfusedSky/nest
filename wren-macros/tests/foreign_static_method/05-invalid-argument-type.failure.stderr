error[E0599]: no function or associated item named `get_slot` found for type `f32` in the current scope
 --> tests/foreign_static_method/05-invalid-argument-type.failure.rs:7:9
  |
7 | fn test(a: f32, b: f32, c: f32) -> f32 {
  |         ^ function or associated item not found in `f32`

error[E0599]: no function or associated item named `get_slot` found for type `f32` in the current scope
 --> tests/foreign_static_method/05-invalid-argument-type.failure.rs:7:17
  |
7 | fn test(a: f32, b: f32, c: f32) -> f32 {
  |                 ^ function or associated item not found in `f32`

error[E0599]: no function or associated item named `get_slot` found for type `f32` in the current scope
 --> tests/foreign_static_method/05-invalid-argument-type.failure.rs:7:25
  |
7 | fn test(a: f32, b: f32, c: f32) -> f32 {
  |                         ^ function or associated item not found in `f32`

error[E0599]: the method `to_output` exists for type `f32`, but its trait bounds were not satisfied
 --> tests/foreign_static_method/05-invalid-argument-type.failure.rs:6:1
  |
6 | #[foreign_static_method]
  | ^^^^^^^^^^^^^^^^^^^^^^^^ method cannot be called on `f32` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `f32: SetValue<'_, wren::context::Foreign>`
          which is required by `f32: ForeignCallOutput`
          `&f32: SetValue<'_, wren::context::Foreign>`
          which is required by `&f32: ForeignCallOutput`
          `&mut f32: SetValue<'_, wren::context::Foreign>`
          which is required by `&mut f32: ForeignCallOutput`
  = note: this error originates in the attribute macro `foreign_static_method` (in Nightly builds, run with -Z macro-backtrace for more info)
